[Style] Convert the 'position-try-fallbacks' property to strong style types

weinig opened the pull request:
#### 607278ffe6a8cff1dbfdae88c0ab45fb7e430e6b
<pre>
[Style] Convert the &apos;position-try-fallbacks&apos; property to strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=302878">https://bugs.webkit.org/show_bug.cgi?id=302878</a>

Reviewed by Darin Adler and Antti Koivisto.

Converts the &apos;position-try-fallbacks&apos; property to use strong style types.

Adds checking of separators to generic list CSSValue -&gt; Style type conversion
to account for the case where the single item optimization kicks in and the
nested type is also a list. In that case, the list must have a different
separator type, so checking at conversion time allows us to select the correct
type to convert to.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/style/PositionTryFallback.cpp: Removed.
* Source/WebCore/rendering/style/PositionTryFallback.h: Removed.
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
* Source/WebCore/style/StyleTreeResolver.h:
* Source/WebCore/style/values/StyleValueTypes+CSSValueConversion.h:
* Source/WebCore/style/values/anchor-position/StylePositionArea.cpp:
* Source/WebCore/style/values/anchor-position/StylePositionTryFallback.cpp: Added.
* Source/WebCore/style/values/anchor-position/StylePositionTryFallback.h: Added.
* Source/WebCore/style/values/anchor-position/StylePositionTryFallbackTactic.h: Added.
* Source/WebCore/style/values/anchor-position/StylePositionTryFallbacks.h: Added.
* Source/WebCore/style/values/images/StyleImageOrientation.cpp:

Canonical link: <a href="https://commits.webkit.org/303580@main">https://commits.webkit.org/303580@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b7b1580e274d841e437ee06c1c57c3d711a8c66

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132893 "9 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5394 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43984 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140428 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84923 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134763 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5778 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5257 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/101617 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68931 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135839 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4057 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119066 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82415 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3944 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83661 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112958 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37185 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143080 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5063 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37768 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/109992 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5145 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4335 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110173 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27928 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3878 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115331 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58605 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5117 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/33685 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4956 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68569 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5207 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5075 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/values/CSSValueAggregates.h:
```diff
@@ -1903,4 +1967,5 @@ struct MarkableTraits<WebCore::CustomIdentifier> {
     static WebCore::CustomIdentifier emptyValue() { return WebCore::CustomIdentifier { nullAtom() }; }
 };
 
+
```
```suggestion
```

darinadler commented on Source/WebCore/style/values/anchor-position/StylePositionTryFallback.cpp:
```diff
@@ -0,0 +1,188 @@
+    if (positionArea.properties && other.positionArea.properties)
+        return positionArea == other.positionArea;
+
+    if (!positionArea.properties && !positionArea.properties)
```
Consider why our tests didn't catch this mistake?
```suggestion
    if (!positionArea.properties && !other.positionArea.properties)
```

darinadler commented on Source/WebCore/style/values/anchor-position/StylePositionTryFallback.cpp:
```diff
@@ -0,0 +1,188 @@
+            return { };
+        }
+
+        auto rule = std::optional<ScopedName> { };
+        auto tactics = SpaceSeparatedVector<PositionTryFallbackTactic> { };
```
Would this be better style? Which do you prefer?
```suggestion
        std::optional<ScopedName> rule;
        SpaceSeparatedVector<PositionTryFallbackTactic> tactics;
```

weinig replied:
I prefer the `auto foo =` one, as seeing the variable name seems more important to me.

darinadler commented on Source/WebCore/style/values/anchor-position/StylePositionTryFallback.cpp:
```diff
@@ -0,0 +1,188 @@
+    ASSERT(positionAreaProperties);
+    RefPtr areaValue = positionAreaProperties->getPropertyCSSValue(CSSPropertyPositionArea);
+    ASSERT(areaValue);
+    return areaValue.releaseNonNull();
```
The `releaseNonNull` function should include the assertion. If it does, then I suggest:
```suggestion
    return positionAreaProperties->getPropertyCSSValue(CSSPropertyPositionArea).releaseNonNull();
```

darinadler commented on Source/WebCore/style/values/anchor-position/StylePositionTryFallback.cpp:
```diff
@@ -0,0 +1,188 @@
+auto CSSValueCreation<PositionTryFallback::PositionArea>::operator()(CSSValuePool&, const RenderStyle&, const PositionTryFallback::PositionArea& value) -> Ref<CSSValue>
+{
+    RefPtr positionAreaProperties = value.properties;
+    ASSERT(positionAreaProperties);
```
I feel like I just asked this, but doesn't the `RefPtr` `->` operator include the assertion that it's not null?

weinig replied:
It does, you are right.

weinig committed:
[Style] Convert the 'position-try-fallbacks' property to strong style types
https://bugs.webkit.org/show_bug.cgi?id=302878

Reviewed by Darin Adler and Antti Koivisto.

Converts the 'position-try-fallbacks' property to use strong style types.

Adds checking of separators to generic list CSSValue -> Style type conversion
to account for the case where the single item optimization kicks in and the
nested type is also a list. In that case, the list must have a different
separator type, so checking at conversion time allows us to select the correct
type to convert to.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/style/PositionTryFallback.cpp: Removed.
* Source/WebCore/rendering/style/PositionTryFallback.h: Removed.
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
* Source/WebCore/style/StyleTreeResolver.h:
* Source/WebCore/style/values/StyleValueTypes+CSSValueConversion.h:
* Source/WebCore/style/values/anchor-position/StylePositionArea.cpp:
* Source/WebCore/style/values/anchor-position/StylePositionTryFallback.cpp: Added.
* Source/WebCore/style/values/anchor-position/StylePositionTryFallback.h: Added.
* Source/WebCore/style/values/anchor-position/StylePositionTryFallbackTactic.h: Added.
* Source/WebCore/style/values/anchor-position/StylePositionTryFallbacks.h: Added.
* Source/WebCore/style/values/images/StyleImageOrientation.cpp:

Canonical link: https://commits.webkit.org/303580@main
